### PR TITLE
Improve `AsyncContent`

### PR DESCRIPTION
**Pull request by lorban**
Improve javadoc and testing + generously comment the surprising implementation.

While I was at it, I also removed one todo.

### CODE REVIEW

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java [hunk 1]
```diff
@@ -49,12 +49,25 @@ public class AsyncContent implements Content.Sink, Content.Source, Closeable
 
     /**
      * {@inheritDoc}
-     * <p>The write completes when the {@link Content.Chunk} returned by {@link #read()}
-     * that wraps {@code byteBuffer} is released.</p>
+     * <p>The write completes:</p>
+     * <ul>
+     * <li>successfully when the {@link Content.Chunk} returned by {@link #read()} is released</li>
+     * <li>successfully just before the {@link Content.Chunk} is returned if the latter {@link Content.Chunk#hasRemaining() has no remaining byte}</li>
+     * </ul>
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java:57]: Worth documenting when the callback is failed.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java [hunk 2]
```diff
@@ -68,6 +81,7 @@ public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
      * write is complete.</p>
      * <p>The callback completes:</p>
      * <ul>
+     * <li>immediately with a failure when this instance is closed or in error</li>
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java]: Please specify you mean "already" in error.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java (hunk 1 above)
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java:57]: Done.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java (hunk 2 above)
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java]: Done.

**Review by sbordet — APPROVED** [review]

**Pull request merged**